core/src/main/java/jenkins/diagnostics/ControllerExecutorsAgents.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2021, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.diagnostics;
+
+import hudson.Extension;
+import hudson.Main;
+import hudson.model.AdministrativeMonitor;
+import jenkins.model.Jenkins;
+import org.jenkinsci.Symbol;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+import org.kohsuke.stapler.interceptor.RequirePOST;
+
+import java.io.IOException;
+
+@Extension
+@Symbol("controllerExecutorsWithAgents")
+@Restricted(NoExternalUse.class)
+public class ControllerExecutorsAgents extends AdministrativeMonitor {
+
+    @Override
+    public String getDisplayName() {
+        return Messages.ControllerExecutorsAgents_DisplayName();
+    }
+
+    @Override
+    public boolean isSecurity() {
+        return true;
+    }
+
+    @Override
+    public boolean isActivated() {
+        return !Main.isDevelopmentMode && Jenkins.get().getNumExecutors() > 0 &&
+                (!Jenkins.get().clouds.isEmpty() || !Jenkins.get().getNodes().isEmpty());
```

Comment:
A new `Nodes#hasNodes()` method could be used for a minor optimization.
Also, caching Jenkins instance in a local variable would help.

Reply:
We could add a method like that, but I'm not sure it would accomplish much. The class naming and usage for nodes in that area is already quite a mess and I don't if adding a utility method to `Nodes` would help much. We could add a utility to `Jenkins` to check for nodes or clouds, but it's a pretty specific, easy check.

I'm not sure that caching the Jenkins instance would help, but I can do that if desired.